### PR TITLE
Pull request for OOP school library: decorate a class

### PR DESCRIPTION
**Pull request by zuhebahmed88091**
Dear reviewer,

You can find the code for OOP school library: decorate a class.

In this PR:
---

### Interface
- [ ] Created a class `Nameable`.
    - Implement a method called `correct_name` that will raise a `NotImplementedError`.

### Turn Person class to Nameable
- [ ] Made sure that  Person class inherits from Nameable.
- [ ] Made sure that this class has a method `correct_name` implemented. It simply return the name attribute.

### Prepare base Decorator

- [ ] Made sure that it inherits from Nameable.
- [ ] In the constructor assigned a nameable object from params to an instance variable.
- [ ] Implemented the `correct_name` method that returns the result of the `correct_name` method of the `@nameable`.

### Prepare CapitalizeDecorator

- [ ] Created a class that inherits from the base Decorator class.
- [ ] Implemented a method `correct_name` that capitalizes the output of `@nameable.correct_name`.

### Prepare TrimmerDecorator

- [ ] Created a class that inherits from the base Decorator class.
- [ ] Implement a method `correct_name` that makes sure that the output of `@nameable.correct_name` has a maximum of 10 characters. If it's longer it should trim the word.

- [ ] See decorators in action:
    - Implemented:
    `person = Person.new(age: 22, name: 'maximilianus')
person.correct_name
capitalized_person = CapitalizeDecorator.new(person)
puts capitalized_person.correct_name
capitalized_trimmed_person = TrimmerDecorator.new(capitalized_person)
puts capitalized_trimmed_person.correct_name`

    For testing purpose in `test.rb`
    
    
testruby

Thank you for reviewing my PR!

Regards,
@zuhebahmed88091

### CODE REVIEW

File: nameable.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Nameable
+  def correct_name
+    raise NotImplementedError, "Subclasses must implement the 'correct_name' method."
+  end
+end
```
Comment (V-Blaze) [inline nameable.rb:5]: - You have done great so far :+1: But there are a few changes you need to make before moving to the next milestone. :smile: :wave: 

- [x] According to the project requirement, you are meant to run this test to see your decorator in action.

```sh
person = Person.new(22, 'maximilianus')
person.correct_name
capitalized_person = CapitalizeDecorator.new(person)
puts capitalized_person.correct_name
capitalized_trimmed_person = TrimmerDecorator.new(capitalized_person)
puts capitalized_trimmed_person.correct_name
```
-  [x] When I ran these, I got an error in the console, I'll suggest you go through your classes and make sure everything is implemented correctly, and then run the code to make sure the result you get is what's expected. :smile: :small_airplane: :wave: 

- - Screenshot for reference:
 
Screenshot from 2023-08-29 11-22-49

**Review by V-Blaze — CHANGES_REQUESTED** [review]
## Hi @zuhebahmed88091 

## Required changes 🖊️ 
# Almost there 🟡 🟦 🟨 

giphy


Great work so far, You have come a long way, keep moving forward👏 👍  
You've done well implementing the project requirements, but there are still some issues I think you need to improve on. 😄 👋 


 ## Highlights
- All required classes created ✔
-  Some class methods implemented correctly 👋 
-  Correct git flow✔

 ## Required Changes ♻️

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person.rb [hunk 2]
```diff
@@ -1,8 +1,11 @@
...
+
+class Person < Nameable
   attr_accessor :name, :age
   attr_reader :id
 
   def initialize(age:, name: 'Unknown', parent_permission: true)
+    super()
     @id = Random.rand(1..1000)
     @name = name
     @age = age
```
Comment (jerryowusu) [inline person.rb]: Good job so far for the effort in making the requested changes by the previous reviewer, however the issue is still not resolved. I do have errors when I run either the person.rb or the new test.rb file you created as shown below. Kindly check that the methods are correctly linked and the Person class is correctly initialized. 
Kindly ensure that it is running well before resubmiting for approval. You can even add screenshots. :pray: 

Screenshot from 2023-08-29 11-07-28

Screenshot from 2023-08-29 11-10-00

**Review by jerryowusu — CHANGES_REQUESTED** [review]
Hi @zuhebahmed88091 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

so-close

 ## Highlights
Great job so far :+1: 
Neat code :+1: 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person.rb (hunk 2 above)
Comment (zuhebahmed88091) [reply person.rb]: I have added screenshots in PR message

Comment (zuhebahmed88091) [reply person.rb]: testruby

Comment (zuhebahmed88091) [reply person.rb]: @jerryowusu  I think you have to add ->
require_relative 'person'
require_relative 'trimmer_decorator'
require_relative 'capitalized_decorator'

for checking the result in person.rb

**Comment by zuhebahmed88091** [comment]
Dear @jerryowusu 

I have already created a `test.rb` file to test as per requirements. As you can see that when I am running the test it is working completely fine. Screenshot is shown for the test result:

testruby

However, as you tried to run a customized test from your end. You forgot to add these line: 

```
require_relative 'person'
require_relative 'trimmer_decorator'
require_relative 'capitalized_decorator'

```
Thus you could not see the accurate result.

File: person.rb (hunk 2 above)
Comment (Abenezer-Tilahun) [reply person.rb]: Fixed.

**Review by Abenezer-Tilahun — APPROVED** [review]
Hi @zuhebahmed88091 
 
Your project is complete  👍 💯   There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 👏🏻

✔️ No linter errors.
✔️ You used correct Git-Flow.
✔️ Professional README  
✔️ Good Commit Messages
✔️ Good job for Implemented Suggested changes.

## Optional suggestions 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

**Pull request merged**